Add missing and remove unused dependencies to add Yarn v3 compatibility

fdel-car opened the pull request:
### What does it do?

- It adds missing dependencies in our packages, and removes unused ones.
- It harmonizes dependencies versions across the repo in order to have everything hoisted up to the root, reducing the size of the cumulated installed packages and also helping us test Strapi [using `portal` protocols](https://github.com/yarnpkg/berry/blob/master/CHANGELOG.md#installs-3), mainly used when linking a project to our codebase. More on that later.

### Why is it needed?

Packages need to be declared explicitly in order for yarn v3 to work properly (for now PnP doesn't work, I'll need to investigate further).
Also fixes almost all the warnings you can get when using yarn v3 about peer dependencies.

#### Details
- Upgraded apollo-server-koa in the graphql plugins to 3.4.0 which is the version from which they fixed their peer dependencies versions by adding a caret to the koa version requested (see [here](https://github.com/apollographql/apollo-server/blob/main/CHANGELOG.md#v340)). Just in case, I also updated apollo-server-core.
- Did not add require-from-string in the helper-plugin dependencies list because it should not have been added as a peer dependency in storybook (see [here](https://github.com/storybookjs/storybook/issues/18241)).
- `react-virtualized` warnings cannot be fixed, their latest release does not provide support in their peer dependencies list for react 17 & 18. We should think about switching to react-window.
- Warnings from `graphql-iso-date` and `graphql-type-long` won't be fixed, those packages are not maintained anymore, we should switch to [`graphql-scalars`](https://github.com/Urigo/graphql-scalars). cc: @Convly

### How to test it?

You can create a strapi project using the below command from this branch after you've run `yarn` once.

```bash
./node_modules/.bin/create-strapi-app --no-run ~/my-strapi-project
```

From this new project, you will need to switch to yarn v3 with `yarn set version stable` then you need to run:

```bash
yarn link --all ~/strapi # Or wherever the branch of strapi associated with this PR is living
```

From there you'll be able to test various `nodeLinker` settings. PnP is still broken - even in `loose` mode - at the time of writing this edit.

### Related issue(s)/PR(s)

A lot of issues and PRs are related to this, I'm gonna update that once this PR is not a draft anymore.

### Remarks

- I'll probably try to revamp later how we handle our tests, in order to run tests in every package locally (directly in their respective folder) we'd need to add `jest` in their `devDependencies`. As mentioned recently, this is a bit painful to maintain, so we'll probably remove the possibility to run tests from packages directly and instead use `--selectProjects` from the root of the repository.

gu-stav commented on package.json:
```diff
@@ -68,51 +68,41 @@
     "prettier": "1.19.1",
     "qs": "6.11.0",
-    "react-test-renderer": "17.0.2",
     "request": "2.88.2",
```
I think we can remove `request` from the list of devDependencies here too ✨

derrickmehaffy replied:
It would be really nice to remove request for good if possible: 
![image](https://user-images.githubusercontent.com/8593673/179076300-a092d6b0-be7b-4662-82e1-b57ae3d0eff0.png)

fdel-car replied:
`request` is required by `request-promise-native` which we are using :/

derrickmehaffy reviewed (commented):
I think we really should decide if we are gonna pin stuff or not because we don't really pick one side or another when it comes to carets

gu-stav commented on packages/core/admin/package.json:
```diff
@@ -46,7 +46,6 @@
-    "@fortawesome/free-brands-svg-icons": "^5.15.3",
     "@fortawesome/free-solid-svg-icons": "^5.15.3",
     "@fortawesome/react-fontawesome": "^0.2.0",
     "@pmmmwh/react-refresh-webpack-plugin": "0.5.7",
```
This one needs an update to `2.0.0`, since jest was upgraded to v28 in https://github.com/strapi/strapi/pull/13740

fdel-car commented on packages/core/admin/package.json:
```diff
@@ -55,26 +54,27 @@
     "font-awesome": "^4.7.0",
-    "formik": "^2.2.6",
-    "fs-extra": "10.0.0",
+    "formik": "^2.2.9",
```
I agree that things are a bit up in the air at the moment, I basically stuck with what was most used in each case - and kept consistency, even when it wouldn't resolve to another version - but I planned on having a chat about this, We should have it now honestly, cc: @alexandrebodin.

It's a complex topic 🙃

Pros of pinning dependencies:
- We basically avoid any surprises if any of our dependencies decide that they want to break semver and introduce breaking changes (or just introduce a bug in a minor a patch release, nobody's perfect).
Keep in mind that we would not be bulletproof anyway as we could still be impacted by issues like this if our dependencies themselves do not pin theirs, this security stops on the first layer that's what I'm saying.

Cons:
- As package authors we want to could want to use wider ranges to prevent duplication in the downloaded modules of our users. Pinning messes up with that. It's definitely less of a problem in a node environment.
- Security fixes released in our dependencies needs to wait for the next release for them to be available - easily - to our users.
- A bit of upgrade noise in the repo related to the PRs upgrading our dependencies.

fdel-car replied:
Unrelated but we could use [that feature of yarn v3](https://yarnpkg.com/features/constraints) also to avoid version differences between our dependencies in the repo.

fdel-car commented:
Too many changes since last time so I decided to start again, it was easier that way and I took this opportunity to write a script to share how I was checking all this.

You can find the `depcheck.ts` (yes in TypeScript cause why not? Easier to work with - I can convert it to JS if you really want to). This script is using the package [`depcheck`](https://github.com/depcheck/depcheck), which allows us to easily see if a package is missing any dependencies or having unused ones.
**This is not 100% accurate**, but it definitely gives good pointers.

Also worth noting, you might notice that I switched to yarn v3 in Strapi itself as well. It's way easier to work with for multiple reasons, so I'll keep it in this PR for now.
I'll revert it before merging.

Last thing I updated the `How to test it?` section in my initial message.

MrChrisRodriguez commented:
This keeps getting pushed -- do y'all have a target release date? We're waiting with bated breath for this to be released so that we can upgrade Strapi (otherwise we've run into a ton of issues with missing dependencies).

fdel-car commented:
Sorry for the delay but I'm doing this in my spare time, since fixing this issue would be helpful to the community I've requested to be able to work on it during work hours soon (in 2 weeks I'll be able to get back to it I think).
Will provide updates as soon as possible.

MrChrisRodriguez commented:
Thanks @fdel-car -- two quick thoughts:
1. I know there were folks in the community willing to help with these improvements but they got shut down. Would their help speed things up here or does it need to be handled by a core contributor?
2. This is a moving target and incremental progress still adds value. Are smaller per-project PRs more likely to get integrated?

SalahAdDin commented:
I got the same problem today, how could I address it?

MrChrisRodriguez commented:
It's wild: multiple devs submitted PRs to try to fix this and were told "no thanks, we'll handle it" then nada.

SalahAdDin commented:
> It's wild: multiple devs submitted PRs to try to fix this and were told "no thanks, we'll handle it" then nada.

So a bad thing.

joshuaellis commented:
@fdel-car not a replacement but this PR will address a lot of the issues people have using yarn3 & pnpm – #15848

joshuaellis commented:
Closing because we merged #15848.
